File: frontend/mobile/lib/core/services/local_storage_service.dart
import 'dart:convert';
import 'dart:io';
import 'package:flutter/foundation.dart';
import 'package:hive_flutter/hive_flutter.dart';
import 'package:path_provider/path_provider.dart';
import 'package:archive/archive.dart';
import 'package:crypto/crypto.dart';

/// Storage priority levels for content
enum StoragePriority {
  critical, // Essential content (Quran text, prayer times)
  high,     // Frequently accessed (bookmarks, recent content)
  medium,   // Cached content (tafsir, hadith)
  low,      // Optional content (images, audio)
}

/// Storage item metadata
class StorageItem {
  final String key;
  final int size;
  final DateTime lastAccessed;
  final DateTime createdAt;
  final StoragePriority priority;
  final bool compressed;
  final String? checksum;

  StorageItem({
    required this.key,
    required this.size,
    required this.lastAccessed,
    required this.createdAt,
    required this.priority,
    this.compressed = false,
    this.checksum,
  });

  Map<String, dynamic> toJson() => {
        'key': key,
        'size': size,
        'lastAccessed': lastAccessed.toIso8601String(),
        'createdAt': createdAt.toIso8601String(),
        'priority': priority.index,
        'compressed': compressed,
        'checksum': checksum,
      };

  factory StorageItem.fromJson(Map<String, dynamic> json) {
    return StorageItem(
      key: json['key'],
      size: json['size'],
      lastAccessed: DateTime.parse(json['lastAccessed']),
      createdAt: DateTime.parse(json['createdAt']),
      priority: StoragePriority.values[json['priority']],
      compressed: json['compressed'] ?? false,
      checksum: json['checksum'],
    );
  }

  StorageItem copyWith({
    DateTime? lastAccessed,
    int? size,
  }) {
    return StorageItem(
      key: key,
      size: size ?? this.size,
      lastAccessed: lastAccessed ?? this.lastAccessed,
      createdAt: createdAt,
      priority: priority,
      compressed: compressed,
      checksum: checksum,
    );
  }
}

/// Storage statistics
class StorageStats {
  final int totalSize;
  final int availableSpace;
  final int usedSpace;
  final int itemCount;
  final Map<StoragePriority, int> sizeByPriority;
  final DateTime lastCleanup;

  StorageStats({
    required this.totalSize,
    required this.availableSpace,
    required this.usedSpace,
    required this.itemCount,
    required this.sizeByPriority,
    required this.lastCleanup,
  });

  double get usagePercentage => (usedSpace / totalSize) * 100;
  bool get isNearCapacity => usagePercentage > 80;
  bool get isCritical => usagePercentage > 95;

  Map<String, dynamic> toJson() => {
        'totalSize': totalSize,
        'availableSpace': availableSpace,
        'usedSpace': usedSpace,
        'itemCount': itemCount,
        'sizeByPriority': sizeByPriority.map((k, v) => MapEntry(k.index.toString(), v)),
        'lastCleanup': lastCleanup.toIso8601String(),
      };
}

/// Download progress callback
typedef DownloadProgressCallback = void Function(int received, int total);

/// Local storage service with smart space management
class LocalStorageService {
  final Box _dataBox;
  final Box _metadataBox;
  
  // Configuration
  static const int maxStorageSize = 500 * 1024 * 1024; // 500MB
  static const int compressionThreshold = 10 * 1024; // 10KB
  static const Duration cleanupInterval = Duration(days: 7);
  static const Duration oldContentThreshold = Duration(days: 30);

  LocalStorageService(this._dataBox, this._metadataBox);

  /// Initialize storage service
  static Future<LocalStorageService> initialize() async {
    await Hive.initFlutter();
    final dataBox = await Hive.openBox('local_storage_data');
    final metadataBox = await Hive.openBox('local_storage_metadata');
    return LocalStorageService(dataBox, metadataBox);
  }

  /// Store data with automatic compression and space management
  Future<void> store(
    String key,
    dynamic data, {
    StoragePriority priority = StoragePriority.medium,
    bool forceCompression = false,
  }) async {
    // Convert data to bytes
    final bytes = _serializeData(data);
    final originalSize = bytes.length;

    // Compress if needed
    bool compressed = false;
    List<int> finalData = bytes;

    if (forceCompression || originalSize > compressionThreshold) {
      finalData = _compressData(bytes);
      compressed = true;
    }

    // Calculate checksum
    final checksum = _calculateChecksum(finalData);

    // Check if we need to free space
    await _ensureSpace(finalData.length, priority);

    // Store data
    await _dataBox.put(key, finalData);

    // Store metadata
    final metadata = StorageItem(
      key: key,
      size: finalData.length,
      lastAccessed: DateTime.now(),
      createdAt: DateTime.now(),
      priority: priority,
      compressed: compressed,
      checksum: checksum,
    );

    await _metadataBox.put(key, jsonEncode(metadata.toJson()));
  }

  /// Retrieve data with automatic decompression
  Future<T?> retrieve<T>(String key) async {
    final data = _dataBox.get(key);
    if (data == null) return null;

    // Update last accessed time
    final metadataJson = _metadataBox.get(key);
    if (metadataJson != null) {
      final metadata = StorageItem.fromJson(jsonDecode(metadataJson));
      final updated = metadata.copyWith(lastAccessed: DateTime.now());
      await _metadataBox.put(key, jsonEncode(updated.toJson()));
    }

    // Decompress if needed
    List<int> bytes = data is List<int> ? data : List<int>.from(data);
    
    final metadataStr = _metadataBox.get(key);
    if (metadataStr != null) {
      final metadata = StorageItem.fromJson(jsonDecode(metadataStr));
      if (metadata.compressed) {
        bytes = _decompressData(bytes);
      }

      // Verify checksum
      if (metadata.checksum != null) {
        final currentChecksum = _calculateChecksum(data is List<int> ? data : List<int>.from(data));
        if (currentChecksum != metadata.checksum) {
          debugPrint('Checksum mismatch for key: $key');
          await remove(key);
          return null;
        }
      }
    }

    return _deserializeData<T>(bytes);
  }

  /// Check if key exists
  bool has(String key) {
    return _dataBox.containsKey(key);
  }

  /// Remove item
  Future<void> remove(String key) async {
    await _dataBox.delete(key);
    await _metadataBox.delete(key);
  }

  /// Get storage statistics
  Future<StorageStats> getStats() async {
    int totalUsed = 0;
    final sizeByPriority = <StoragePriority, int>{
      StoragePriority.critical: 0,
      StoragePriority.high: 0,
      StoragePriority.medium: 0,
      StoragePriority.low: 0,
    };

    for (var key in _metadataBox.keys) {
      final metadataJson = _metadataBox.get(key);
      if (metadataJson != null) {
        final metadata = StorageItem.fromJson(jsonDecode(metadataJson));
        totalUsed += metadata.size;
        sizeByPriority[metadata.priority] = 
            (sizeByPriority[metadata.priority] ?? 0) + metadata.size;
      }
    }

    final lastCleanup = await _getLastCleanupTime();

    return StorageStats(
      totalSize: maxStorageSize,
      availableSpace: maxStorageSize - totalUsed,
      usedSpace: totalUsed,
      itemCount: _dataBox.length,
      sizeByPriority: sizeByPriority,
      lastCleanup: lastCleanup,
    );
  }

  /// Download content with progress tracking
  Future<void> downloadContent(
    String key,
    Future<List<int>> Function() downloader, {
    StoragePriority priority = StoragePriority.medium,
    DownloadProgressCallback? onProgress,
  }) async {
    try {
      final data = await downloader();
      
      if (onProgress != null) {
        onProgress(data.length, data.length);
      }

      await store(key, data, priority: priority);
    } catch (e) {
      debugPrint('Download failed for $key: $e');
      rethrow;
    }
  }

  /// Smart cleanup - removes old and low-priority content
  Future<void> performCleanup({bool force = false}) async {
    final stats = await getStats();
    final lastCleanup = stats.lastCleanup;
    
    if (!force && DateTime.now().difference(lastCleanup) < cleanupInterval) {
      return; // Too soon for cleanup
    }

    final items = <StorageItem>[];
    for (var key in _metadataBox.keys) {
      final metadataJson = _metadataBox.get(key);
      if (metadataJson != null) {
        items.add(StorageItem.fromJson(jsonDecode(metadataJson)));
      }
    }

    // Sort by priority (low first) and age (old first)
    items.sort((a, b) {
      // Critical items are never removed
      if (a.priority == StoragePriority.critical) return 1;
      if (b.priority == StoragePriority.critical) return -1;

      // Compare by priority first
      final priorityCompare = a.priority.index.compareTo(b.priority.index);
      if (priorityCompare != 0) return -priorityCompare;

      // Then by age
      return a.lastAccessed.compareTo(b.lastAccessed);
    });

    // Remove old content
    final now = DateTime.now();
    int freedSpace = 0;
    
    for (var item in items) {
      if (item.priority == StoragePriority.critical) continue;
      
      final age = now.difference(item.lastAccessed);
      final shouldRemove = age > oldContentThreshold || 
                          (stats.isNearCapacity && item.priority == StoragePriority.low);

      if (shouldRemove) {
        await remove(item.key);
        freedSpace += item.size;
        
        // Stop if we've freed enough space
        if (!stats.isCritical && freedSpace > maxStorageSize * 0.2) {
          break;
        }
      }
    }

    await _setLastCleanupTime(DateTime.now());
    debugPrint('Cleanup completed. Freed ${freedSpace ~/ 1024}KB');
  }

  /// Ensure enough space is available
  Future<void> _ensureSpace(int requiredSize, StoragePriority priority) async {
    final stats = await getStats();
    
    if (stats.availableSpace >= requiredSize) {
      return; // Enough space available
    }

    // Need to free space
    final items = <StorageItem>[];
    for (var key in _metadataBox.keys) {
      final metadataJson = _metadataBox.get(key);
      if (metadataJson != null) {
        final item = StorageItem.fromJson(jsonDecode(metadataJson));
        // Only consider items with lower priority
        if (item.priority.index > priority.index) {
          items.add(item);
        }
      }
    }

    // Sort by priority (low first) and age (old first)
    items.sort((a, b) {
      final priorityCompare = b.priority.index.compareTo(a.priority.index);
      if (priorityCompare != 0) return priorityCompare;
      return a.lastAccessed.compareTo(b.lastAccessed);
    });

    int freedSpace = 0;
    for (var item in items) {
      await remove(item.key);
      freedSpace += item.size;
      
      if (freedSpace >= requiredSize) {
        break;
      }
    }

    if (freedSpace < requiredSize) {
      throw Exception('Unable to free enough space for storage');
    }
  }

  /// Compress data using GZip
  List<int> _compressData(List<int> data) {
    final encoder = GZipEncoder();
    return encoder.encode(data) ?? data;
  }

  /// Decompress data
  List<int> _decompressData(List<int> data) {
    final decoder = GZipDecoder();
    return decoder.decodeBytes(data);
  }

  /// Calculate checksum
  String _calculateChecksum(List<int> data) {
    return sha256.convert(data).toString();
  }

  /// Serialize data to bytes
  List<int> _serializeData(dynamic data) {
    if (data is List<int>) return data;
    if (data is String) return utf8.encode(data);
    return utf8.encode(jsonEncode(data));
  }

  /// Deserialize data from bytes
  T? _deserializeData<T>(List<int> bytes) {
    if (T == List<int>) return bytes as T;
    
    final str = utf8.decode(bytes);
    if (T == String) return str as T;
    
    try {
      return jsonDecode(str) as T;
    } catch (e) {
      return str as T;
    }
  }

  /// Get last cleanup time
  Future<DateTime> _getLastCleanupTime() async {
    final timestamp = _metadataBox.get('_last_cleanup');
    if (timestamp == null) return DateTime(2000);
    return DateTime.parse(timestamp);
  }

  /// Set last cleanup time
  Future<void> _setLastCleanupTime(DateTime time) async {
    await _metadataBox.put('_last_cleanup', time.toIso8601String());
  }

  /// Clear all storage
  Future<void> clearAll() async {
    await _dataBox.clear();
    await _metadataBox.clear();
  }

  /// Export storage data for backup
  Future<Map<String, dynamic>> exportData() async {
    final data = <String, dynamic>{};
    for (var key in _dataBox.keys) {
      data[key.toString()] = _dataBox.get(key);
    }
    return data;
  }

  /// Import storage data from backup
  Future<void> importData(Map<String, dynamic> data) async {
    for (var entry in data.entries) {
      await _dataBox.put(entry.key, entry.value);
    }
  }

  /// Get device storage info
  Future<Map<String, int>> getDeviceStorageInfo() async {
    try {
      final directory = await getApplicationDocumentsDirectory();
      final stat = await directory.stat();
      
      // This is approximate - actual implementation would need platform-specific code
      return {
        'total': 1024 * 1024 * 1024, // 1GB placeholder
        'free': 512 * 1024 * 1024,   // 512MB placeholder
        'used': 512 * 1024 * 1024,   // 512MB placeholder
      };
    } catch (e) {
      return {
        'total': 0,
        'free': 0,
        'used': 0,
      };
    }
  }
}

File: frontend/nextjs-app/src/lib/services/local-storage-service.ts
import pako from 'pako';

/**
 * Storage priority levels for content
 */
export enum StoragePriority {
  CRITICAL = 0, // Essential content (Quran text, prayer times)
  HIGH = 1,     // Frequently accessed (bookmarks, recent content)
  MEDIUM = 2,   // Cached content (tafsir, hadith)
  LOW = 3,      // Optional content (images, audio)
}

/**
 * Storage item metadata
 */
export interface StorageItem {
  key: string;
  size: number;
  lastAccessed: Date;
  createdAt: Date;
  priority: StoragePriority;
  compressed: boolean;
  checksum?: string;
}

/**
 * Storage statistics
 */
export interface StorageStats {
  totalSize: number;
  availableSpace: number;
  usedSpace: number;
  itemCount: number;
  sizeByPriority: Record<StoragePriority, number>;
  lastCleanup: Date;
}

/**
 * Local storage service with smart space management for web
 */
export class LocalStorageService {
  private static readonly MAX_STORAGE_SIZE = 500 * 1024 * 1024; // 500MB
  private static readonly COMPRESSION_THRESHOLD = 10 * 1024; // 10KB
  private static readonly CLEANUP_INTERVAL = 7 * 24 * 60 * 60 * 1000; // 7 days
  private static readonly OLD_CONTENT_THRESHOLD = 30 * 24 * 60 * 60 * 1000; // 30 days

  private static readonly METADATA_PREFIX = '_metadata_';
  private static readonly DATA_PREFIX = '_data_';
  private static readonly LAST_CLEANUP_KEY = '_last_cleanup';

  /**
   * Store data with automatic compression and space management
   */
  static async store(
    key: string,
    data: any,
    priority: StoragePriority = StoragePriority.MEDIUM,
    forceCompression: boolean = false
  ): Promise<void> {
    try {
      // Serialize data
      const serialized = JSON.stringify(data);
      const bytes = new TextEncoder().encode(serialized);
      const originalSize = bytes.length;

      // Compress if needed
      let compressed = false;
      let finalData: Uint8Array = bytes;

      if (forceCompression || originalSize > this.COMPRESSION_THRESHOLD) {
        finalData = pako.gzip(bytes);
        compressed = true;
      }

      // Calculate checksum
      const checksum = await this.calculateChecksum(finalData);

      // Check if we need to free space
      await this.ensureSpace(finalData.length, priority);

      // Store data in IndexedDB
      await this.storeInIndexedDB(this.DATA_PREFIX + key, finalData);

      // Store metadata
      const metadata: StorageItem = {
        key,
        size: finalData.length,
        lastAccessed: new Date(),
        createdAt: new Date(),
        priority,
        compressed,
        checksum,
      };

      localStorage.setItem(
        this.METADATA_PREFIX + key,
        JSON.stringify(metadata)
      );
    } catch (error) {
      console.error('Failed to store data:', error);
      throw error;
    }
  }

  /**
   * Retrieve data with automatic decompression
   */
  static async retrieve<T>(key: string): Promise<T | null> {
    try {
      // Get metadata
      const metadataStr = localStorage.getItem(this.METADATA_PREFIX + key);
      if (!metadataStr) return null;

      const metadata: StorageItem = JSON.parse(metadataStr);

      // Update last accessed time
      metadata.lastAccessed = new Date();
      localStorage.setItem(
        this.METADATA_PREFIX + key,
        JSON.stringify(metadata)
      );

      // Get data from IndexedDB
      const data = await this.retrieveFromIndexedDB(this.DATA_PREFIX + key);
      if (!data) return null;

      // Verify checksum
      if (metadata.checksum) {
        const currentChecksum = await this.calculateChecksum(data);
        if (currentChecksum !== metadata.checksum) {
          console.warn('Checksum mismatch for key:', key);
          await this.remove(key);
          return null;
        }
      }

      // Decompress if needed
      let bytes = data;
      if (metadata.compressed) {
        bytes = pako.ungzip(data);
      }

      // Deserialize
      const text = new TextDecoder().decode(bytes);
      return JSON.parse(text) as T;
    } catch (error) {
      console.error('Failed to retrieve data:', error);
      return null;
    }
  }

  /**
   * Check if key exists
   */
  static has(key: string): boolean {
    return localStorage.getItem(this.METADATA_PREFIX + key) !== null;
  }

  /**
   * Remove item
   */
  static async remove(key: string): Promise<void> {
    localStorage.removeItem(this.METADATA_PREFIX + key);
    await this.removeFromIndexedDB(this.DATA_PREFIX + key);
  }

  /**
   * Get storage statistics
   */
  static async getStats(): Promise<StorageStats> {
    let totalUsed = 0;
    const sizeByPriority: Record<StoragePriority, number> = {
      [StoragePriority.CRITICAL]: 0,
      [StoragePriority.HIGH]: 0,
      [StoragePriority.MEDIUM]: 0,
      [StoragePriority.LOW]: 0,
    };

    let itemCount = 0;

    // Iterate through all metadata
    for (let i = 0; i < localStorage.length; i++) {
      const key = localStorage.key(i);
      if (key?.startsWith(this.METADATA_PREFIX)) {
        const metadataStr = localStorage.getItem(key);
        if (metadataStr) {
          const metadata: StorageItem = JSON.parse(metadataStr);
          totalUsed += metadata.size;
          sizeByPriority[metadata.priority] += metadata.size;
          itemCount++;
        }
      }
    }

    const lastCleanupStr = localStorage.getItem(this.LAST_CLEANUP_KEY);
    const lastCleanup = lastCleanupStr
      ? new Date(lastCleanupStr)
      : new Date(2000, 0, 1);

    return {
      totalSize: this.MAX_STORAGE_SIZE,
      availableSpace: this.MAX_STORAGE_SIZE - totalUsed,
      usedSpace: totalUsed,
      itemCount,
      sizeByPriority,
      lastCleanup,
    };
  }

  /**
   * Smart cleanup - removes old and low-priority content
   */
  static async performCleanup(force: boolean = false): Promise<void> {
    const stats = await this.getStats();
    const now = Date.now();

    if (!force && now - stats.lastCleanup.getTime() < this.CLEANUP_INTERVAL) {
      return; // Too soon for cleanup
    }

    const items: StorageItem[] = [];

    // Collect all items
    for (let i = 0; i < localStorage.length; i++) {
      const key = localStorage.key(i);
      if (key?.startsWith(this.METADATA_PREFIX)) {
        const metadataStr = localStorage.getItem(key);
        if (metadataStr) {
          items.push(JSON.parse(metadataStr));
        }
      }
    }

    // Sort by priority (low first) and age (old first)
    items.sort((a, b) => {
      // Critical items are never removed
      if (a.priority === StoragePriority.CRITICAL) return 1;
      if (b.priority === StoragePriority.CRITICAL) return -1;

      // Compare by priority first
      const priorityCompare = b.priority - a.priority;
      if (priorityCompare !== 0) return priorityCompare;

      // Then by age
      return (
        new Date(a.lastAccessed).getTime() -
        new Date(b.lastAccessed).getTime()
      );
    });

    // Remove old content
    let freedSpace = 0;
    const usagePercentage = (stats.usedSpace / stats.totalSize) * 100;

    for (const item of items) {
      if (item.priority === StoragePriority.CRITICAL) continue;

      const age = now - new Date(item.lastAccessed).getTime();
      const shouldRemove =
        age > this.OLD_CONTENT_THRESHOLD ||
        (usagePercentage > 80 && item.priority === StoragePriority.LOW);

      if (shouldRemove) {
        await this.remove(item.key);
        freedSpace += item.size;

        // Stop if we've freed enough space
        if (usagePercentage < 95 && freedSpace > this.MAX_STORAGE_SIZE * 0.2) {
          break;
        }
      }
    }

    localStorage.setItem(this.LAST_CLEANUP_KEY, new Date().toISOString());
    console.log(`Cleanup completed. Freed ${(freedSpace / 1024).toFixed(0)}KB`);
  }

  /**
   * Ensure enough space is available
   */
  private static async ensureSpace(
    requiredSize: number,
    priority: StoragePriority
  ): Promise<void> {
    const stats = await this.getStats();

    if (stats.availableSpace >= requiredSize) {
      return; // Enough space available
    }

    // Need to free space
    const items: StorageItem[] = [];

    for (let i = 0; i < localStorage.length; i++) {
      const key = localStorage.key(i);
      if (key?.startsWith(this.METADATA_PREFIX)) {
        const metadataStr = localStorage.getItem(key);
        if (metadataStr) {
          const item: StorageItem = JSON.parse(metadataStr);
          // Only consider items with lower priority
          if (item.priority > priority) {
            items.push(item);
          }
        }
      }
    }

    // Sort by priority (low first) and age (old first)
    items.sort((a, b) => {
      const priorityCompare = b.priority - a.priority;
      if (priorityCompare !== 0) return priorityCompare;
      return (
        new Date(a.lastAccessed).getTime() -
        new Date(b.lastAccessed).getTime()
      );
    });

    let freedSpace = 0;
    for (const item of items) {
      await this.remove(item.key);
      freedSpace += item.size;

      if (freedSpace >= requiredSize) {
        break;
      }
    }

    if (freedSpace < requiredSize) {
      throw new Error('Unable to free enough space for storage');
    }
  }

  /**
   * Calculate checksum using SubtleCrypto
   */
  private static async calculateChecksum(data: Uint8Array): Promise<string> {
    const hashBuffer = await crypto.subtle.digest('SHA-256', data);
    const hashArray = Array.from(new Uint8Array(hashBuffer));
    return hashArray.map((b) => b.toString(16).padStart(2, '0')).join('');
  }

  /**
   * Store data in IndexedDB
   */
  private static async storeInIndexedDB(
    key: string,
    data: Uint8Array
  ): Promise<void> {
    return new Promise((resolve, reject) => {
      const request = indexedDB.open('SanadStorage', 1);

      request.onerror = () => reject(request.error);

      request.onupgradeneeded = (event) => {
        const db = (event.target as IDBOpenDBRequest).result;
        if (!db.objectStoreNames.contains('data')) {
          db.createObjectStore('data');
        }
      };

      request.onsuccess = () => {
        const db = request.result;
        const transaction = db.transaction(['data'], 'readwrite');
        const store = transaction.objectStore('data');
        const putRequest = store.put(data, key);

        putRequest.onsuccess = () => resolve();
        putRequest.onerror = () => reject(putRequest.error);
      };
    });
  }

  /**
   * Retrieve data from IndexedDB
   */
  private static async retrieveFromIndexedDB(
    key: string
  ): Promise<Uint8Array | null> {
    return new Promise((resolve, reject) => {
      const request = indexedDB.open('SanadStorage', 1);

      request.onerror = () => reject(request.error);

      request.onupgradeneeded = (event) => {
        const db = (event.target as IDBOpenDBRequest).result;
        if (!db.objectStoreNames.contains('data')) {
          db.createObjectStore('data');
        }
      };

      request.onsuccess = () => {
        const db = request.result;
        const transaction = db.transaction(['data'], 'readonly');
        const store = transaction.objectStore('data');
        const getRequest = store.get(key);

        getRequest.onsuccess = () => resolve(getRequest.result || null);
        getRequest.onerror = () => reject(getRequest.error);
      };
    });
  }

  /**
   * Remove data from IndexedDB
   */
  private static async removeFromIndexedDB(key: string): Promise<void> {
    return new Promise((resolve, reject) => {
      const request = indexedDB.open('SanadStorage', 1);

      request.onerror = () => reject(request.error);

      request.onsuccess = () => {
        const db = request.result;
        const transaction = db.transaction(['data'], 'readwrite');
        const store = transaction.objectStore('data');
        const deleteRequest = store.delete(key);

        deleteRequest.onsuccess = () => resolve();
        deleteRequest.onerror = () => reject(deleteRequest.error);
      };
    });
  }

  /**
   * Clear all storage
   */
  static async clearAll(): Promise<void> {
    // Clear localStorage metadata
    const keysToRemove: string[] = [];
    for (let i = 0; i < localStorage.length; i++) {
      const key = localStorage.key(i);
      if (key?.startsWith(this.METADATA_PREFIX) || key?.startsWith(this.DATA_PREFIX)) {
        keysToRemove.push(key);
      }
    }
    keysToRemove.forEach((key) => localStorage.removeItem(key));

    // Clear IndexedDB
    return new Promise((resolve, reject) => {
      const request = indexedDB.deleteDatabase('SanadStorage');
      request.onsuccess = () => resolve();
      request.onerror = () => reject(request.error);
    });
  }
}
